abd-clam/src/core/dataset/vec2d.rs

```diff
@@ -50,6 +56,163 @@ impl<T: Send + Sync + Copy, U: Number> VecDataset<T, U> {
+            .write_all(&dimensionality_bytes)
+            .map_err(|err| format!("{err}"))?;
+
+        // Write individual vectors
+        for row in &self.data {
+            for column in *row {
+                let column_bytes = column.to_le_bytes();
+                handle.write_all(&column_bytes).map_err(|err| format!("{err}"))?;
+            }
+        }
+
+        // If the reordering map exists, write it and preprend a 1 for the flag byte. Otherwise write
+        // a 0 byte
+        if let Some(map) = &self.reordering {
+            // Write one byte corresponding to if the reordering map exists
+            handle.write(&[1_u8]).map_err(|err| format!("{err}"))?;
+
+            for idx in map {
+                let fullwidth = *idx as u64;
+
+                handle
+                    .write_all(&fullwidth.to_le_bytes())
+                    .map_err(|err| format!("{err}"))?;
+            }
+        } else {
+            handle.write(&[0]).map_err(|err| format!("{err}"))?;
+        }
+
+        Ok(())
+    }
+
+    /// Reads a saved numeric dataset from a specified path. The buffer passed in is the actual place where data is stored
+    /// because `Vec2d` only holds references
+    ///
+    /// # Errors
+    /// Errors on read write issues
+    pub fn load(
+        path: &Path,
+        metric: fn(&[V], &[V]) -> U,
+        is_expensive: bool,
+        buffer: &'a mut Vec<Vec<V>>,
+    ) -> Result<Self, String> {
+        let mut handle = File::open(path).map_err(|err| format!("{err}"))?;
+
+        // Decode the header (Assert the header is equal to 'VEC2D')
+        let mut header_buf = [0_u8; 5];
+        handle.read_exact(&mut header_buf).map_err(|err| format!("{err}"))?;
+        let found: String = header_buf.map(|c| c as char).iter().collect();
+        // If the header didn't match, return an error
+        if found != "VEC2D" {
+            return Err("Invalid header".to_string());
+        }
+
+        // Get the dataset's name
+        let mut name = String::new();
+        let mut char_buf = [0];
+        handle.read_exact(&mut char_buf).map_err(|err| format!("{err}"))?;
+
+        while char_buf[0] != 0 {
+            name.push(char_buf[0] as char);
+            handle.read_exact(&mut char_buf).map_err(|err| format!("{err}"))?;
+        }
+
+        // Get the cardinality and dimensionality
+        let mut cardinality_buf = [0_u8; 8];
+        let mut dimensionality_buf = [0_u8; 8];
+
+        // Read in the cardinality and dimensionality
+        handle
+            .read_exact(&mut cardinality_buf)
+            .map_err(|err| format!("{err}"))?;
+        handle
+            .read_exact(&mut dimensionality_buf)
+            .map_err(|err| format!("{err}"))?;
+
+        // Convert them to usize
+        let cardinality = usize::from_le_bytes(cardinality_buf);
+        let dimensionality = usize::from_le_bytes(dimensionality_buf);
+
+        // Get each data point
+        let type_size = std::mem::size_of::<V>();
```

Comment:
Change this in accordance with the `type_name` that you will have written. Ideally, you would check to see that the name matches and then use the `num_bytes` method on `Number`.